### ISSUE TITLE
Existance of Solutions Assumptions

### ISSUE THREAD

**Issue by jlombs** [uwhpsc-2016/homework2#3]
For the triangular matrix part of the homework, should we assume that the input matrices aren't poorly scaled, and moreover that a solution exists? I didn't see anything in the writeup, but I'm wondering if we have to mimic the "LinAlgError" behavior of the numpy solver as well, testing and filtering cases in the event that we have singular matrices etc...

**Comment by mvelegar** [comment]
Yes, assume that a solution exists. All the test matrices we'll be using for this assignment will be "well-behaved" and SDD, so for the purposes of this exercise you need not worry about either singular or badly-scaled matrices.

**Comment by jlombs** [comment]
Great, thanks!

**Comment by cswiercz** [comment]
That's a good observation. In fact, with a triangular matrix it's actually quite easy to check for existence of solutions. Can you bring your linear algebra skills to the table and find an easy to compute check?

**Comment by jlombs** [comment]
>  Can you bring your linear algebra skills to the table and find an easy to compute check?

Well you can check for full rank, which is an O(n^3) computation in general but for a triangular matrix, you can just check for any non-zero entries on the diagonal (since det(A)=prod(A_{ii})), but that may not solve all your issues depending on what you consider a solution.

If you want a *unique* solution, you're definitely good. 

But say my lower-triangular matrix is really a (n-1)x(n-1) sub-matrix, with the first diagonal element being a 0 and the rest of the matrix being full rank. This is rank deficient, but if the image vector you are asking for also has a 0 in the first element, then there is an infinite family of solutions (the usual algorithm would fail unless it tested for this case though). This can happen for other elements too if there are zeros in other places in the image vector, or if the image vector has a very strict structure. 

So if we are just concerned with finding any vector x such that Ax = b, we need a coordinated check with the image vector as well. You'd have to do something like projecting your whole problem it into its minimal span, checking if the span of your matrix matched the image vector, solving it in the reduced space, and then transforming back and padding the redundant rows with 0's (to provide one representative element of solution class of vectors). A little more costly that just looking for 0's on the diagonal!
